File: 0083-remove-duplicates-from-sorted-list/0083-remove-duplicates-from-sorted-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        visit = defaultdict(bool)
        l, r = ListNode(), head
        l.next = head
        while r != None:
            if not visit[r.val]:
                l = l.next
                l.val = r.val
                visit[r.val] = True
            r = r.next
        
        l.next = None
        
        return head

File: 0083-remove-duplicates-from-sorted-list/0083-remove-duplicates-from-sorted-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null) return null;
        
        ListNode s = head; ListNode f = head;
        
        while (f != null) {
            while (f != null && s.val == f.val) f = f.next;
            if (f != null) {
                s = s.next;
                s.val = f.val;
            }
        }
        
        s.next = null;
        return head;
    }
}